### ISSUE TITLE
Technical discovery into revised "time to complete" forms metrics

### ISSUE THREAD

**Issue by andreahewitt-odd** [department-of-veterans-affairs/va.gov-team#31789]
## Issue Description
Form completion rate and form time to complete -- We received a recent flag about this from Rachael because Anna thinks it's being calculated by session but it's really being captured by multiple sessions. 
---
## Tasks
- [x] How are these currently being calculated in Domo 
- [x] Identify if there is work to do in DOMO
- [x] Identify if there is work to do in BigQuery

## Acceptance Criteria
- [ ] Technical discovery is complete and tickets have been created

**Comment by chloedotbrown** [comment]
### Form completion rate

As mentioned in [recap](https://vfs.atlassian.net/wiki/spaces/AT/pages/1934393412/Data+changes+for+KPI+dashboards) in Confluence, some VFS engineers have expressed concern about this metric being calculated incorrectly because it differs from what they see in other sources. For example, where the Domo shows a 38% rate for a form, one engineer stated that the rate was actually 92%. They hypothesized that the reason for the difference was that the rate was being calculated by session instead of by user, meaning that longer forms taking more sessions to complete would have falsely low rates.

**Findings:**
It is true that the current [Beast Mode](https://va-gov.domo.com/datacenter/beastmode?id=1025) in Domo for this KPI is calculated by session. This would be easy enough to change. But while calculating by user does increase the completion rate, **it does NOT explain the large difference in completion rates across data sources:**
![Form Completion Methodology](https://user-images.githubusercontent.com/45518242/140580235-26859dff-d10d-4af8-a973-436b6621020b.png)
The completion rate by user was also consistent with user counts at the start and end of the funnel visualizations, which are built off a different standardized Forms view. Just to be safe, I did some further exploration to see if the regex in the Beast Mode's denominator was capturing more sessions/users than intended and accidentally deflating the completion rate. However, I found no evidence that this was the case. It seems like this would be unlikely, given how the views are constructed. 

**Next steps:**
I think we need more information about how this metric is being calculated in Grafana / wherever else people are looking. I have to say, a success rate of 92% sounds very high to me for people doing anything on the internet. 

It is possible that this is actually evidence of a communication problem, rather than a technical one. But if so, it's still one worth addressing, as I think it may be causing distrust in the forms dashboard in particular, which does still have low update among VFS teams.

### Average time to complete
By comparison, this metric seems like a pretty easy ticket to scope. As [discussed](https://vfs.atlassian.net/wiki/spaces/AT/pages/1934393412/Data+changes+for+KPI+dashboards), the Domo ETL _does_ total time spent on form by session, which is going to deflate time to complete for any user who takes more than one session. 

<img src="https://user-images.githubusercontent.com/45518242/140585260-ec23eff9-5a33-42ae-a75f-a46f8cce2834.png" width=600>

This one could probably be fixed pretty easily in Domo – it would just mean removing `session` from the aggregation, plus potentially adding a step to account for users who may have successfully submitted the same form type more than once. Not sure how frequently that happens, but certainly seems possible / worth accounting for.

CC: @bsmartin-ep @obrienjulie

**Comment by chloedotbrown** [comment]
**From 12/2 meeting with Jonathan, Julie, Brian, and Chloe**

**_Form completion rate:_**
Have updated the version of Beast Mode used in the dashboard and data dictionary definition to be scoped by user. We feel confident that this is correct. Next steps for this metric will be education/communication efforts to address discrepancy between our completion rate vs. the API success rate developers see on Grafana, which they don't always understand as different. 

_Suggestions:_ present brief explainer in ToT (possibly part of series of misunderstood metrics, differences in data sources, part of user research, etc), build out data dictionary to call out metrics we know are misunderstood, etc.

**_Avg. time to complete_**
Cannot be resolved in ETL, need to address the large number of null or 0 `time_on_step` values in [Standard Forms Funnel](https://va-gov.domo.com/datasources/432a85a7-5581-4e80-8601-dc13c9654436/details/overview) data. For next steps, Brian and Julie will experiment with alternate method for calculating `time_on_step` used in forms_engagement view. If this doesn't resolve the issue, may explore other techniques.

**Comment by obrienjulie** [comment]
From 12/9 meeting with Jonathan, Brian, Chloe and Julie

The BQ views were leveraging a case statement that was limiting the results which resulted in the large number of null or 0 results.  

The BQ views have been updated to remove the case statement for both the hit_time and next_hit_time.

**Comment by chloedotbrown** [comment]
Since we're still working on this and don't want teams looking at incorrect information, I've removed both "avg. minutes to complete" cards from the Forms dashboard ([Highlights](https://va-gov.domo.com/page/447193050/kpis/details/1632380729) and [Ease of Use](https://va-gov.domo.com/page/447193050/kpis/details/204304216) versions) and replaced them with notecards letting viewers know they are under construction. 

Once we've resolved calculation issues, we can return to dashboard an update #vfs-analytics channel.

**Comment by michelle-dooley** [comment]
Meeting on 2/15 with Michelle. Brian, Julie & Chloe

@bsmartin-ep  to complete updated metrics as discussed and compare to see if in line with expectations.  Target completion by end Sprint 72 (3/1)

**Comment by chloedotbrown** [comment]
## Time-to-complete QA
### Proposed acceptance criteria
Since we don't have numbers to check this against, I'd recommend we shoot for something like:
- [x] Transformations match the problem we are trying to solve.
- [ ] Patterns in output data is consistent with definitions of custom metrics.
- [ ] Data is within a range that makes sense for what we know of how users interact with the product.
- [ ] Relationships between new metrics match expectations - or assumptions can be updated and still be consistent with the business context.
- [ ] Outliers are minimal and a threshold for filtering can be easily determined.

### Problems
We're really close! I think there are still some weird things in the data, but hopefully they are easy to solve. 

First, we're still getting really low totals for both hit and step approaches for calculating time-to-complete, which aren't fixed by filtering for a minimum number of steps completed. Even filtering for a minimum of 3 steps, there are still users with extremely low totals. Second, there are too many instances where `total_hit_time` > `total_step_time`, which seems like it shouldn't be happening, based on the definitions in [Brian's original Slack message](https://adhoc.slack.com/archives/C01EUNU6W1Z/p1647360892240539):
> `Hit Time` is the (perhaps poorly named) difference between a page and the absolute next page, regardless if it is in our form lookup table.  So closer to the traditional time on page metric.
> `Step Time` (which is what we mostly use now) is the difference between a page and the next page in the lookup table.

I'd expect `total_step_time` to always be equal to or greater than `total_hit_time`. For example, if a user spent 5 minutes on all form steps, spend 4 minutes on the VA homepage, and then returned to review and submit in 1 minute, their `total_hit_time` would be only 6 minutes but the `total_step_time` would be 10. But that's often not the case in this data, and it affects some forms more than others:

**Hit time is higher than step time (by user completion)**
<img width="450" alt="Screen Shot 2022-03-18 at 4 28 34 PM" src="https://user-images.githubusercontent.com/45518242/159096635-36c24778-2ee4-4a0d-8bb4-f8d17e936bcf.png">

 ### QA process
<img width="350" alt="QA dashboard" align="right" src="https://user-images.githubusercontent.com/45518242/159103718-1bde06a5-7304-441d-9196-9587c0df75af.png">

I put together a [rough dashboard](https://va-gov.domo.com/page/1751626899) to help myself understand the data and see if filters affected distributions the way I expected them to. Everyone should have access to this - though warning that because Domo's histogram offerings are so awful, I had to bootleg the ones here with bar charts so they shouldn't be used outside of the 1 hr or so range (scatter plots do a better job of showing full spread, though they're not as helpful for understanding lower totals). In addition to looking at the impact of filters on measures of center + spread, I spent a lot of time looking at the step vs. hit time comparison at the bottom.

Here were the basic steps I went through:
- Looked at the base data set and transformation code
- Looked at averages + medians by form
- Looked at distribution of both `total_step_time` and `total_hit_time` by form
- Looked at relationship between `total_step_time` and `total_hit_time` for each user form completion
- Put everything on a temporary dashboard and tested impact of adding filters to weed out potential data from reloads of confirmation page:
  - Minimum `steps` - tried 2, 3, and 4
  - Minmum `total_step_time` - 1 and 2 minutes
  - Minimum `total_hit_time` - 1 and 2 minutes
  - ^plus combinations of above settings
- Spot checked user journeys for outliers
  - Grabbed `user` for people with big discrepancies between their `total_step_time` and `total_hit_time`
  - Filtered base dataset by `user`, `session` and sorted by `visitstarttime`, `hit_number`
    - User ID: 8473901018439139665 | Session: 8.473901018439139E28
    - User ID: 4547969458683558415 | Session: 4.547969458683558E28

### Possible issues
It looks like since the `session` field of the base dataset is coming in as a number, Domo is shortening it to scientific notation without retaining the full value, which is clumping adjacent sessions into just one. With the few test user journeys I did where hit time was higher than step time, I would see the same `session` with multiple `visitstarttime` values. That said - `session` isn't part of the partition for any of the window functions, so I'm at a loss how this would be part of the problem! Could be a red herring?

### Additional data
Will we also be able to use one of these approaches to calculate **average time on step**? Past versions of this metric on the dashboard included this to compare with exit rates and would be nice to be able to do again. This metric doesn't need to be limited to only folks who complete the form, so it seems _possible_ we might be able to do it with the [new, non-cross-joined base dataset](https://va-gov.domo.com/datasources/02048e78-cdf4-4488-a026-0e2dba7801e8/details/settings) if `Hit Time` was selected?

### Next steps
@bsmartin-ep , would love to get your thoughts on this since you've been in the weeds with it for awhile! I can also continue to do some poking around on the dashboard and in the base data next week, if that seems useful.

**Comment by bsmartin-ep** [comment]
@chloedotbrown / @obrienjulie / @jonathan-epstein13  -

Primary Redshift ETL for time-to-complete and time-on-step is updated:
[[Redshift] Form Events and Calculated Metrics
](https://va-gov.domo.com/datacenter/dataflows/177/details#data-lineage)

### Output Datasets

[[OUTPUT] Forms Engagement - Time on Step ](https://va-gov.domo.com/datasources/9251be6b-fc95-417b-a971-24024a559033/details/overview)
[[OUTPUT] Forms Engagement - Time to Complete 
](https://va-gov.domo.com/datasources/d8ac9e92-5488-442a-84fc-bc84af3e5880/details/overview)

The only restrictions being applied:

- All hits must be within 30 days of the confirmation timestamp (set in both output ETLs):

```sql
AND ("closest_confirmation_timestamp" - "hit_timestamp") BETWEEN 0 AND 2592000
```

- Next step time is NOT calculated for the Confirmation step.  I think 1. we don't want to count that towards the SUM and 2. we were getting extra large values due to users restarting the same form in the same session.

```sql
    CASE
        WHEN "step_name" != 'Confirmation' THEN
        LEAD("visit_start_time" + ("hit_time" / 1000), 1) OVER (
            PARTITION BY "user",
            "visit_start_time",
            "form_name"
            ORDER BY
                "hit_number" ASC
        )
    END as next_step_timestamp,
```
### Mitigating outliers

We still have a lot of low time-to-completion values.  Especially for the 526EZ form, which I believe is due to a coding error that was fixed a few months ago.  Otherwise, some of these are a natural effect of splitting applications over several GA user IDs.

My suggestion would be to exclude any confirmation timestamps that are less than 3 steps or less than 2 (maybe 3?) minutes while calculating time-to-complete.

**Comment by chloedotbrown** [comment]
Thanks @bsmartin-ep - this is great!

### Mitigating outliers
Yes! Adding the requirements of 3 or more steps and at least 3 min to complete across all forms sounds good to me. I imagine these conditions would filter out most of the 526EZ error data, and we could add a note or reach out to the team to let them know that they'll see fewer forms in the time metrics during that period for this reason. @michelle-dooley - does that sound good?

### Matching output data
Looking at the two cards that will go on the Forms dashboard (currently at top of [temporary dashboard](https://va-gov.domo.com/page/1751626899)), there are still some differences in total time between the two datasources when you filter by form. I think this will be confusing to users – it's confusing to me! – and I can't think of any reason in the business logic why these totals would not match. But looking at the script for the output datasets, there's nothing that jumps out at me as the obvious reason. Any thoughts?

#### Average minutes to complete comparisons
| Form        | [Time to complete (total)](https://va-gov.domo.com/page/1751626899/kpis/details/633749245)          | [Time on step (running total)](https://va-gov.domo.com/page/1751626899/kpis/details/240479125)  |
| :------------- |-------------:| -----:|
| Form 10-10CG - Caregivers      | 6.7 | 24.8 |
| Form 10-10EZ - Health Benefits      | 24.2 | 32.7 |
| Form 20-0996 - Higher Level Review     | 8.7 | 8.1 |
| Form 526EZ - Disability Claims     | 22.6 | 44.1 |
| Form 20-0996 - Higher Level Review     | 8.7 | 8.1 |
| Form 22-1995 - Request for Change    | 12.8 | 13.1 |
| Form 28-1900 - Chapter 31     | 1.5 | 6.1 |
| Form 5655 - Financial Status Report     | 22.0 | 43.6 |

**Comment by michelle-dooley** [comment]
@chloedotbrown @obrienjulie @bsmartin-ep - yes I am good with the 3 or more steps requirements for mitigating outliers and just  note it so everyone is aware of the threshold for those metrics.

**Comment by bsmartin-ep** [comment]
@chloedotbrown / @obrienjulie -

As discussed in the meeting today, I think we see this discrepancy between time-to-complete and the running time-on-step totals for two reasons...

1. We have two different date fields.  Time-to-complete is based on a date extracted from the closest_confirmation timestamp.  All hits are associated with that timestamp regardless of when they actually happened.  Time-on-step is the date of the actual hit we are examining.  Which one to use?  Would it be more confusing to have them differ or valuable to see the true hit date?  Perhaps to evaluate the affect of changes to the site on individual step times?

(Side note - from what I read in the Domo help, these UTC timestamps should be automatically converted to our configured time zone.  This may be worth testing.)

2. Our time-to-complete is an average of SUMs vs the running average which is a SUM of averages.  I don't think these will usually match.

I created a test output dataset and ETL to re-SUM the individual time-on-step values and average them.  They're very close! I think the minor discrepancy can be explained as rounding errors:

https://va-gov.domo.com/kpis/details/1850969821

**Comment by chloedotbrown** [comment]
**Decisions reached in 5/17 meeting**
- Do not think second concern raised above is driving discrepancy between totals and running totals of steps. I put together a [proof of concept](https://docs.google.com/spreadsheets/d/1esW24my9xcEYFCWRIyVTPciMnlIJ7347I8RLj758u6M/edit?usp=sharing) spreadsheet showing that both approaches return the same total for fictional user data. So long as both datasets include the same rows and grouping logic, this should not be the issue.
- We _do_ think that the first date issue may be causing these issues. @obrienjulie will be editing the time-by-step dataflow to make sure `closest_confirmation_timestamp` is being used, instead of the `date` when the step occurred. I will test after.
- We will focus our edits on the two initial data outputs created by the Redshift dataflow:
  - [[Output] Forms Engagement - Time to complete](https://va-gov.domo.com/datasources/d8ac9e92-5488-442a-84fc-bc84af3e5880/details/overview)
  - [[Output] Forms Engagement - Time on step](https://va-gov.domo.com/datasources/9251be6b-fc95-417b-a971-24024a559033/details/overview)

**Comment by obrienjulie** [comment]
The primary differences between the Forms dashboard and the Time to Complete are the filters we have added. That seems to be the primary reason for dropping records between both. 

In the TTC output, we are using this filter:

 WHERE
     "closest_confirmation_timestamp" IS NOT NULL
     AND ("closest_confirmation_timestamp" - "hit_timestamp") BETWEEN 0 AND 2592000 (0-30 Days).

I do not want to remove either of these as then we would get too much data that just wouldn’t make sense for the TTC metric. I did play around with the 0-30 day filter, upping it to 45 days but that didn’t move the needle too much.   For now, I think we keep it at the 0-30 day range.   

I did add in the step_name to the TOS output and it’s running now.  The confirmation_timestamp has also been added to both datasets as a date field, labelled date.  

I have the ETL running now and it usually takes ~4 hours to finish.

**Comment by chloedotbrown** [comment]
Hi @obrienjulie - think we had a miscommunication about the edits to the additional dataflow edits! I went into the testing dashboard, and the totals that had previously matched were way off (screenshot below).

The `step_name` field only needed to be added to the "Time on Step" output dataset. Adding it to the fields and grouping conditions for the "Time to Complete" dataset means that these sums still get calculated by step and no longer give us an estimate of the total time spent on the process. 

Similarly, I was hoping to _cast_ the `closest_confirmation_timestamp` field as a datetime data type, not add in the existing `date` field. Re-introducing this field to the output and grouping conditions confounds the decision we'd previously for the business logic to assign all steps to the date the form was completed, not the date they occurred.

I restored the Redshift ETL to the previous version, which had your corrections for adding `step_name` to the 2nd dataset and standardized the browser/device fields. I did try a few strategies to cast the confirmation date myself, but they weren't supported by Domo's version of Redshift - imagine this limitation is probably why Brian created the conversion beast mode in the first place. I'll solve our filter issue by just renaming those as "date." 

Datasets are re-running now and I'll double-check them on Monday.
